# Chloe deBeus

## 01. Import

In [1]:
import pandas as pd
import numpy as np
import os

## 02. Data wrangling from reading

In [2]:
mainPath = r'C:\Users\18286\Documents\Career Foundry\05_2023_Instacart Basket Analysis'

In [3]:
mainPath

'C:\\Users\\18286\\Documents\\Career Foundry\\05_2023_Instacart Basket Analysis'

In [4]:
df_orders = pd.read_csv(os.path.join(mainPath, '02 Datasets', 'Original data', 'orders.csv'), index_col = False)

In [5]:
df_orders.shape

(3421083, 7)

In [6]:
#looking at columns
df_orders.columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'order_dow',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [7]:
#looking at types
df_orders.dtypes

order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [8]:
#looking at table generally
df_orders.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [9]:
#renaming the column because of confusion of title of variable
df_orders.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

In [10]:
#checking the proper change was made
df_orders.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [11]:
#next step to understand table better
df_orders.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02       6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [12]:
df_orders['order_id'] = df_orders['order_id'].astype('str')

In [13]:
df_orders.dtypes

order_id                   object
user_id                     int64
eval_set                   object
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [14]:
#dropping column
df_orders = df_orders.drop(columns = ['eval_set'])

In [16]:
df_orders.columns

Index(['order_id', 'user_id', 'order_number', 'order_day_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [18]:
#bring in products csv and generally explore
df_products = pd.read_csv(os.path.join(mainPath, '02 Datasets', 'Original data', 'products.csv'), index_col = False)

In [19]:
df_products.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices'], dtype='object')

In [20]:
df_products.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [21]:
df_products.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [22]:
#bring in department dataset and explore it
df_depts = pd.read_csv(os.path.join(mainPath, '02 Datasets', 'Original data', 'departments.csv'), index_col = False)

In [23]:
df_depts.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [24]:
df_depts.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [25]:
#transcribing the table, so it goes from horizontal to vertical
df_depts_T = df_depts.T

In [27]:
df_depts_T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [26]:
df_depts_T.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [28]:
new_header = df_depts_T.iloc[0]

In [29]:
new_header

0    department
Name: department_id, dtype: object

In [30]:
df_depts_T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [31]:
df_depts_T = df_depts_T[1:]

In [32]:
df_depts_T_new = df_depts_T

In [33]:
df_depts_T_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [34]:
df_depts_T_new.reset_index()

index                0
0      1           frozen
1      2            other
2      3           bakery
3      4          produce
4      5          alcohol
5      6    international
6      7        beverages
7      8             pets
8      9  dry goods pasta
9     10             bulk
10    11    personal care
11    12     meat seafood
12    13           pantry
13    14        breakfast
14    15     canned goods
15    16       dairy eggs
16    17        household
17    18           babies
18    19           snacks
19    20             deli
20    21          missing

In [44]:
df_depts_T_new.columns = new_header

In [45]:
df_depts_T_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [48]:
data_dict_depts = df_depts_T_new.to_dict('index')

In [49]:
data_dict_depts

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [50]:
df_products.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [51]:
print(data_dict_depts.get('19'))

{'department': 'snacks'}


In [53]:
snacks = df_products[df_products['department_id'] == 19]                          

In [54]:
snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

## 03. Further wrangling with 4.4 Task

### Step two

In [58]:
#change other variable to suitable format
#looking for said varaible
df_orders.describe()

user_id  order_number  order_day_of_week  order_hour_of_day  \
count  3.421083e+06  3.421083e+06       3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01       2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01       2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00       0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00       1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01       3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01       5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02       6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [62]:
df_orders.dtypes

order_id                   object
user_id                     int64
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [63]:
#don't need user_id as an int
df_orders['user_id'] = df_orders['user_id'].astype('str')

In [65]:
#now changed data types
df_orders.dtypes

order_id                   object
user_id                    object
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

### Step Three

In [66]:
df_orders.columns
# when looking, it's a bit confusing that there is an order id and order number. what's the difference?

Index(['order_id', 'user_id', 'order_number', 'order_day_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [69]:
df_orders

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0        2539329       1             1                  2                  8   
1        2398795       1             2                  3                  7   
2         473747       1             3                  3                 12   
3        2254736       1             4                  4                  7   
4         431534       1             5                  4                 15   
...          ...     ...           ...                ...                ...   
3421078  2266710  206209            10                  5                 18   
3421079  1854736  206209            11                  4                 10   
3421080   626363  206209            12                  1                 12   
3421081  2977660  206209            13                  1                 12   
3421082   272231  206209            14                  6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [68]:
df_orders['order_number'].value_counts(dropna = False)

1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: order_number, Length: 100, dtype: int64

In [70]:
df_orders_edit_one = df_orders

In [71]:
df_orders_edit_one.rename(columns = {'order_number':'customer_order_count'}, inplace = True)

In [72]:
df_orders_edit_one

order_id user_id  customer_order_count  order_day_of_week  \
0        2539329       1                     1                  2   
1        2398795       1                     2                  3   
2         473747       1                     3                  3   
3        2254736       1                     4                  4   
4         431534       1                     5                  4   
...          ...     ...                   ...                ...   
3421078  2266710  206209                    10                  5   
3421079  1854736  206209                    11                  4   
3421080   626363  206209                    12                  1   
3421081  2977660  206209                    13                  1   
3421082   272231  206209                    14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

### Step Four

In [73]:
df_orders['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

#### The busiest hour for orders is 10AM (Timezone uncertain)

### Step Five

In [87]:
#using the dictionary created earlier from the departments dataframe
data_dict_depts.get('4')

{'department': 'produce'}

#### Deparment four is produce

### Step Six

In [93]:
#print the dictionary again to see which department is breakfast
data_dict_depts

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [101]:
#department 14 is breakfast
breakfast = df_products[df_products['department_id'] == 14]   

In [103]:
#printing subset to ensure it's what we want
breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

### Step Seven

In [104]:
#searching for which we need
data_dict_depts

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [110]:
#5, 20, 12, 7
#I need to use loc function in order to collect multiple ids
dinner_parties = df_products.loc[df_products['department_id'].isin([5, 20, 12, 7])]  

In [112]:
dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

### Step Eight

In [115]:
dinner_parties.shape

(7650, 5)

#### Dinner parties has 7650 rows and 5 columns

### Step Nine

In [122]:
#we need to use the orders dataframe in order to address customers
df_orders

order_id user_id  customer_order_count  order_day_of_week  \
0        2539329       1                     1                  2   
1        2398795       1                     2                  3   
2         473747       1                     3                  3   
3        2254736       1                     4                  4   
4         431534       1                     5                  4   
...          ...     ...                   ...                ...   
3421078  2266710  206209                    10                  5   
3421079  1854736  206209                    11                  4   
3421080   626363  206209                    12                  1   
3421081  2977660  206209                    13                  1   
3421082   272231  206209                    14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [136]:
df_orders['user_id']

0               1
1               1
2               1
3               1
4               1
            ...  
3421078    206209
3421079    206209
3421080    206209
3421081    206209
3421082    206209
Name: user_id, Length: 3421083, dtype: object

In [ ]:
df_orders[]

In [140]:
# remember that we changed this variable to a string, so we have to write '1' as a string not an integer
user_one = df_orders.loc[df_orders['user_id'] == '1']

In [139]:
user_one

order_id user_id  customer_order_count  order_day_of_week  \
0   2539329       1                     1                  2   
1   2398795       1                     2                  3   
2    473747       1                     3                  3   
3   2254736       1                     4                  4   
4    431534       1                     5                  4   
5   3367565       1                     6                  2   
6    550135       1                     7                  1   
7   3108588       1                     8                  1   
8   2295261       1                     9                  1   
9   2550362       1                    10                  4   
10  1187899       1                    11                  4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

### Step Ten

In [141]:
user_one.describe()

customer_order_count  order_day_of_week  order_hour_of_day  \
count             11.000000          11.000000          11.000000   
mean               6.000000           2.636364          10.090909   
std                3.316625           1.286291           3.477198   
min                1.000000           1.000000           7.000000   
25%                3.500000           1.500000           7.500000   
50%                6.000000           3.000000           8.000000   
75%                8.500000           4.000000          13.000000   
max               11.000000           4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

### Step Twelve

In [146]:
df_orders.to_csv(os.path.join(mainPath, '02 Datasets', 'Clean Data', 'orders_wrangled.csv'))

### Step Thirteen

In [147]:
df_depts_T_new.to_csv(os.path.join(mainPath, '02 Datasets', 'Clean Data', 'departments_wrangled.csv'))